<a href="https://colab.research.google.com/github/karsarobert/LLM2026/blob/main/ChatGpt2024_04_feladatokkal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Chat GPT és más nagy nyelvi modellek alkalmazása
##PTE Gépi tanulás
###4. gyakorlat: Ismerkedés a promptolással API-n keresztül
2026. február 23.

**Felfedezések a strukturálatlan szövegekben – Mesterséges intelligencia technikák szövegelemzéshez**

A strukturálatlan szöveges adatok rendkívül értékes felismeréseket rejtenek, azonban a minőségi információk ilyen léptékű értelmezése komoly kihívást jelent. Ezen az órán azokat a technikákat vizsgáljuk, amelyek lehetővé teszik a generatív AI-rendszerek számára a szabad formátumú szövegek automatikus elemzését és rendszerezését.

Megtanuljuk, hogyan osztályozza az érzelem-elemzés (sentiment analysis) a szavak mögötti érzelmi tónust, hogy meghatározza, a tartalom pozitív, semleges vagy negatív-e. Láthatjuk, hogy az adatosztályozás hogyan használja a gépi tanulást a szöveg előre meghatározott témákba és címkékbe történő besorolására. Felfedezzük, hogyan tudják a mintázat-illesztési technikák kinyerni a legfontosabb információkat a strukturálatlan adatokból, és azokat strukturált kimenetekké alakítani.

Ezek az AI-képességek segítik a vállalkozásokat és szervezeteket abban, hogy hatékonyan dolgozzanak fel hatalmas mennyiségű strukturálatlan adatot, és tárjanak fel megvalósítható felismeréseket. Az érzelemanalízis feltárja, hogyan éreznek a fogyasztók a termékekkel vagy márkákkal kapcsolatban. Az adatosztályozás automatikusan rendezi a tartalmakat, például az ügyfélszolgálati jegyeket, témákba. A mintázatfelismerés adatokat nyer ki a dokumentumokból, például a számlákból, és azokat tiszta, strukturált nyilvántartásokká alakítja.

Ez a gyakorlati eszköztár lehetővé teszi bárki számára, hogy hozzáférjen a naponta létrejövő szöveges adatokban rejlő tudáshoz. Vizsgáljuk meg, hogyan segíthet a generatív AI értelmezni a nyelvet nagyobb léptékben.

**A gyakorlaton tárgyalt témák:**

* Érzelem-elemzés – AI technikák az érzelmek felismerésére a szövegben
* Strukturálatlan adatok rendszerezése – hogyan automatizálja az AI a szöveges kategorizálást
* Rossz minőségű adatok tisztítása – hogyan azonosítja és oldja meg az AI a problémákat az adathalmazokban
* Strukturálatlan adatok értelmezése – mintázatfelismerés az információ kinyeréséhez


**Érzelem-elemzés – Mesterséges intelligencia technikák az érzelmek felismerésére a szövegben**

Az érzelem-elemzés (sentiment analysis) az egyik legfontosabb, mesterséges intelligencia technikákat alkalmazó megoldás a  prompt engineering (a természetes nyelvre épülő, mesterséges intelligenciával vezérelt feladatok automatizálásának) területén. Ez az első az általunk tárgyalt AI-alkalmazások sorában. Az érzelem-elemzés természetes nyelvi feldolgozást (Natural Language Processing, NLP) használ az érzelmi jelek azonosítására és kinyerésére a szöveges adatokból. A cél az, hogy felmérjük egy írásmű általános érzelmét, például meghatározzuk, hogy egy termék értékelése pozitív vagy negatív-e.

**Az érzelem-elemzés leggyakoribb felhasználási területei a következők:**

* **Márkamonitorozás:** A márkák nyomon követhetik az online zajló beszélgetéseket, hogy figyelemmel kísérjék hírnevüket. Az érzelem-elemzés lehetővé teszi számukra, hogy felismerjék a felmerülő PR-problémákat, és mérjék a marketingkampányok hatását.
* **Ügyfélszolgálat:** Az ügyfélélményért felelős csapatok nagyszabásúan összesíthetik a visszajelzéseket felmérések, vélemények és közösségi média említések elemzésével. Ez adatvezérelt betekintést nyújt abba, hogy az emberek hol elégedettek vagy csalódottak a termékekkel és szolgáltatásokkal.
* **Piackutatás:** Az ügyfelek hangulatának megértésével a vállalkozások felfedezhetik szolgáltatásaik vagy termékeik hiányosságait, és ennek megfelelően pótolhatják azokat.
* **Politikai elemzés:** A politikai csoportok szemmel tarthatják a választók hozzáállását a jelöltekkel és a problémákkal kapcsolatos viták nyomon követésével. A technológia segít az üzenetek tesztelésében és azoknak a területeknek a meghatározásában, ahol további felvilágosításra van szükség.
* **Közösségi média megfigyelése:** A vállalatok az érzelem-elemzést használják fel a közösségi média bejegyzéseinek hangulatának azonosítására, hogy megértsék a szélesebb fogyasztói magatartást és trendeket.
* **Befektetési elemzés:** A befektetők érzelem-elemzést használhatnak a piaci mozgások előrejelzéséhez. A releváns hírek, fórumok és blogok feldolgozásával az eszközök meg tudják határozni, hogy a befektetők optimisták vagy pesszimisták-e bizonyos részvényekkel vagy szektorokkal kapcsolatban.
* **Közegészségügy:** Az egészségügyi hatóságok érzelem-elemzést alkalmazhatnak a közvélemény aggályainak, a félretájékoztatásnak és az érzelmi állapotoknak a megértésére olyan válságok idején, mint például a világjárványok. Ez az információ segít az erőforrások kommunikációs stratégiáinak és elosztásának irányításában.


#Példa
Prompt: Azonosítsd és kategorizáld a [szövegben] kifejezett érzelmeket (pozitív, semleges, negatív).
szöveg = "Ez a termék csodálatos! Tényleg javított az életemen."

Kimenet:
A szövegben kifejezett érzelmek a következők:

Pozitív: "csodálatos", "javított az életemen"
Ez azt jelenti, hogy az író pozitív érzelmeket fejez ki a termékkel kapcsolatban, kifejezve az elégedettségét és a boldogságát annak hatásaival kapcsolatban.

#Feladat

Módosítsuk az elöbbi promptot úgy, hogy csak 1 szavas kimenetet adjon

#Megoldás

Azonosítsd és kategorizáld a [Text] kifejezett érzelmeket (pozitív, semleges, negatív). Csak egy szóban válaszolj!
Text = "Ez a termék csodálatos! Tényleg javított az életemen."

Kimenet:
Pozitív

#Feladat:
Osztályozzuk az albbi mondatokat az elöbbiek szerint a kimenetetként pedig kérjünk táblázatot!

Szöveg: "Nagyon csalódott vagyok a szörnyű ügyfélszolgálat miatt, amit kaptam. Teljesen eszköztelenek és udvariatlanok voltak."
"Az ételek ebben az étteremben abszolút finomak. Mindenképpen vissza fogok térni még egyszer."
"Nemrég vettem ezt a laptopot, és ez a legjobb technikai vásárlás, amit évek óta csináltam. Annyira gyors és megbízható."
"Úgy néz ki, hogy ma délután esni fog. A biztonság kedvéért vinnem kéne egy esernyőt."
"A projektjavaslatom beadási határideje jövő pénteken 17 órakor lejár."
"Most néztem meg az új Marvel filmet, és teljesen felülmúlta az elvárásaimat. Nagyon szórakoztató volt az elejétől a végéig."
"Ha manapság a híreket olvasom, csak szorongok és lehangoltnak érzem magam a világ helyzete miatt."
"Ez a legrosszabb ízű étel, amit valaha ettem egy étteremben. A steakem túlsült, a zöldségek pedig pépesek voltak."
"Megbántam, hogy a legújabb telefonmodellt vettem meg. Az akkumulátor élettartama szörnyű és tele van hibákkal. Meg kellett volna tartanom a régi telefonomat."
"Nem hiszem el, hogy hetekig tartó tanulás után megbuktam a vizsgámon. Az egész felkészülés a semmiért volt."
"Még ma be kell ugranom az élelmiszerboltba tejért és tojásért."
"Az utasítások szerint a készülék alaphelyzetbe állításához 5 másodpercig nyomva kell tartani a bekapcsológombot."

#Megoldás


Prompt:



| Szöveg                                                                                                           | Hangulat  |
|------------------------------------------------------------------------------------------------------------------|-----------|
| "Nagyon csalódott vagyok a szörnyű ügyfélszolgálat miatt, amit kaptam. Teljesen eszköztelenek és udvariatlanok voltak." | Negatív   |
| "Az ételek ebben az étteremben abszolút finomak. Mindenképpen vissza fogok térni még egyszer."                     | Pozitív   |
| "Nemrég vettem ezt a laptopot, és ez a legjobb technikai vásárlás, amit évek óta csináltam. Annyira gyors és megbízható." | Pozitív   |
| "Úgy néz ki, hogy ma délután esni fog. A biztonság kedvéért vinnem kéne egy esernyőt."                             | Semleges  |
| "A projektjavaslatom beadási határideje jövő pénteken 17 órakor lejár."                                            | Semleges  |
| "Most néztem meg az új Marvel filmet, és teljesen felülmúlta az elvárásaimat. Nagyon szórakoztató volt az elejétől a végéig." | Pozitív   |
| "Ha manapság a híreket olvasom, csak szorongok és lehangoltnak érzem magam a világ helyzete miatt."               | Negatív   |
| "Ez a legrosszabb ízű étel, amit valaha ettem egy étteremben. A steakem túlsült, a zöldségek pedig pépesek voltak." | Negatív   |
| "Megbántam, hogy a legújabb telefonmodellt vettem meg. Az akkumulátor élettartama szörnyű és tele van hibákkal. Meg kellett volna tartanom a régi telefonomat." | Negatív   |
| "Nem hiszem el, hogy hetekig tartó tanulás után megbuktam a vizsgámon. Az egész felkészülés a semmiért volt."      | Negatív   |
| "Még ma be kell ugranom az élelmiszerboltba tejért és tojásért."                                                 | Semleges  |
| "Az utasítások szerint a készülék alaphelyzetbe állításához 5 másodpercig nyomva kell tartani a bekapcsológombot." | Semleges  |

**Megkérheti az LLM-et, hogy állítsa elő a kimenetet Markdown, HTML, HTML táblázat, XML, JSON vagy akár saját minta formájában.**

Az AI képes megérteni és elemezni a strukturálatlan szöveget az adatok pontos osztályozása érdekében.
Az AI bizonyos esetekben előzetes példák nélkül is értelmezni tudja a kategorizálási mintákat.
Tipikus felhasználási módok közé tartoznak például:
* Emailek spamszűrése
* Hírcikkek téma szerinti szortírozása
* Ügyfelek szegmentálása
* Banki hitelkérelmek kockázatának meghatározása

Az AI használatának előnyei:

* Időmegtakarítás az automatizálás által
* Pontosabb és következetesebb eredmények készítése, mint a manuális módszerek.
* Komplex adathalmazok jobb kezelése és elemzése.

#Feladat

Osztályozzuk az emaileket spam/nem spam és rendeljünk hozzájuk egy fontossági pontszámot 1-5 iga ahol az 5 a legfontosabb. Készítsünk táblázatot az eredményből.

emailek:
Tárgy: Itt igényelheti INGYENES ÜDÜLÉSI UTALVÁNYÁT! Feladó: unknownsender@freedeals.com Email tartalma: Kattintson ide , hogy INGYEN élvezhesse a luxus nyaralást! Igyekezzen!Az ajánlatok, mint ez, nem sokáig elérhetőek!"

Tárgy: Q1 2023 Teljesítményértékelés Feladó: HR_dept@yourcompany.com Email tartalma: Kedves János, Jövő héten lesznek az első negyedéves teljesítményértékelések. Kérjük, keddig frissítse a feladatait és eredményeit a dolgozói portálon. Üdvözlettel, HR-csapat

Tárgy: Új termékünk bemutatása Feladó: newsletter@electronicsstore.com Email tartalma: Kedves János, Örömmel tájékoztatjuk Önt régóta várt okostévénk érkezéséről. Nézze meg legjobb ajánlatainkat. Kellemes vásárlást! Üdvözlettel, Elektronikai Bolt

Tárgy: Reunión del proyecto mañana a las 3 PM Feladó: gerente@tuempresa.com Email tartalma: Equipo, La reunión del proyecto está programada para mañana a las 3 PM en la sala de conferencias del quinto piso. Estén preparados para brindar actualizaciones sobre sus tareas. Atentamente, Gerente del Proyecto (magyarul: Csapat, A projekt megbeszélés holnap 3 órára van tervezve az 5. emeleti tárgyalóteremben. Legyenek felkészülve feladataik frissítésével kapcsolatban. Üdvözlettel, Projektvezető)

#Megoldás

| E-mail tárgysora                           | Feladó                      | Fontosság |
|------------------------------------------|--------------------------|-----------|
| Itt igényelheti INGYENES ÜDÜLÉSI UTALVÁNYÁT! | unknownsender@freedeals.com | Spam      |
| Q1 2023 Teljesítményértékelés            | HR_dept@yourcompany.com   | 3         |
| Új termékünk bemutatása                   | newsletter@electronicsstore.com | 4         |
| Reunión del proyecto mañana a las 3 PM   | gerente@tuempresa.com     | 2         |

#Adattisztítás - avagy hogyan azonosít és orvosol az AI problémákat az adathalmazokban

Az adattisztítás egy olyan folyamat, amely során beazonosítják és korrigálják (vagy eltávolítják) a hibákat, eltéréseket és pontatlanságokat egy adathalmazból. Ez magában foglalhatja a duplikált rekordok észlelését, hiányos vagy hiányzó adatok kezelését, értékek ellenőrzését és javítását, az adatok formázását, szabványosítását vagy normalizálását, valamint a kiugró értékek kezelését.

Az adattisztítás fő célja az adatok minőségének és megbízhatóságának javítása, annak biztosítása, hogy azok pontosak, konzisztensek és a felhasználási célra alkalmas formátumúak legyenek. Ez azért döntő fontosságú, mert a hiányos, rendetlen (angolul: *dirty*) adatok zavarhatják az elemzéseket, pontatlan felismerésekhez és következtetésekhez vezethetnek, és negatív hatással lehetnek a döntéshozatali folyamatokra.

Fontos megjegyezni, hogy az adattisztításnak az adatkezelés rendszeres részének kell lennie – nem egyszeri feladat, mivel új hibák keletkezhetnek, amikor új adatok kerülnek hozzáadásra vagy a meglévő adatok módosulnak.

Általánosságban elmondható, hogy az adattisztítás összetett és költséges folyamat volt, amelyet generatív AI-val hatékonyan lehetséges elvégezni.

**Összefoglalás**

Példaként tegyük fel, hogy adatokat gyűjtöttél (lekapartál) a LinkedIn-ről. Megvannak potenciális érdeklődők nevei, és különböző forrásokból megvannak az e-maileik, de előfordulhat, hogy helytelenül vannak formázva. Előfordulhat, hogy az egyik név emojit, a másik pedig olyan attribútumokat (pl. Dr., PhD, vagy MBA) tartalmaz, amikre nincs szükség. Ebben az esetben meg kell tisztítani az adatokat, hogy e-mail kampányokat tudjunk küldeni. A cél az, hogy visszakapjuk az adatokat címek, emojik vagy egyéb karakterek nélkül.


#Példa

Te a szöveges adatok tisztításának szakértője vagy.
Tisztísd meg a [nevek és e-mail címek] listáját. A nevekből távolítsd el az emojikat és a címeket, például orvos, PhD stb. Győződj meg arról is, hogy az e-mailek helyesen vannak formázva.
A javított listáából szészíts egy táblázatot.
nevek és e-mailek:

Kimenet:

| Név               | E-mail              |
|-------------------|---------------------|
| John Appleseed    | john.applecom       |
| Emma Doe          | emma@doe.com       |
| William Smith     | will.smith@uni.com |
| Peter Parker      | peteparker@gmail.com |
| Steve America     | captain@america.com |
| Jane Light        | jane.light@mail.com |
| Sophie Turner     | sophie@example.com |
| Chris Evans       | chris@uni.edu       |

#Feladat

Az alábbi struktúrálatlan szövegből az előzőhöz hasonlóan nyerjük ki az információkat egy táblázatba!

A gépi tanulás területén folyamatosan zajlanak a kutatások és a fejlődések, amelyek mindennapi életünkre is hatással vannak. Például, a gépi tanulás segít meghatározni, mit nézzen meg a Netflix-en a felhasználó, illetve ajánlani tud olyan termékeket a Amazonnak, amelyek megfelelnek a vevő igényeinek. De nem csupán üzleti haszonnal bír a technológia; orvostudományban is alkalmazzák, például rákbetegségek diagnosztizálására.

Dr. János Kovács, a Budapesti Egyetem informatika professzora, egy nagyszerű cikkben bemutatta a gépi tanulás alapjait a "Gépitanulás Magazinnak". A cikkben, amelyet 2023. március 15-én jelentettek meg, a neuronhálózatok, a mélytanulás és a renormalizációs grupák alapelveiről olvashattunk. Ha több információt szeretne kapni erről a térről, akkor ne felejtse el írni az e-mail címére: [janos.kovacs@uni-budapest.hu](mailto:janos.kovacs@uni-budapest.hu) .

Szemléltessék még Dr. Zsuzsanna Nagy cikkét, amely az "Önkéntes Gépi Tanulás Etnikai Előítéletekről" címet viselte. A cikk 2023. február 28-án jelent meg a magazinban. Dr. Nagy elemzi benne, hogyan befolyásolhatják a gépi tanulási modelleket a társadalmi előítéletek, és miért lehet fontos az adathalmozgatás a javításukhoz. További információkat kaphat erre vonatkozóan az e-mail címére: [zsuzsanna.nagy@itm.gov.hu](mailto:zsuzsanna.nagy@itm.gov.hu).

Végezetül, ha érdeklődik az autonóm járművek iránt, akkor Tamás Varga cikke is remek forrást jelent számára. A "Miért Fontos a Gépi Tanulás Az Autonóm Járművek Fejlesztésében?" című cikke 2023. február 14-én került kiadványba. Tamás feltárta benne, hogy miért van szükség a gépi tanulásra az autonóm járművek fejlesztésénél, és milyen új lehetőségeket nyithat ki ennek alkalmazása az iparágban. Ellenőrizze a cikkötleteit Tamás e-mail címére: [tamas.varga@autonomousvehicles.com](mailto:tamas.varga@autonomousvehicles.com).

Biztos, hogy sokoldalú és izgalmas információkat talál ezekben a cikkekben, amelyek nagyban hozzájárulnak a gépi tanulás világának jobb megismeréséhez. Ne felejtse el figyelemmel kísérni a "Gépitanulás Magazin" aktuális kiadványait, és maradj informálva a legfrissebb trendekről és fejlesztésekről!

#Megoldás



Prompt:


**Amint látható, ha több száz vagy akár több ezer olyan rekorddal rendelkezik, amelyeket meg kell tisztítania, ezzel sok időt takaríthat meg. Előfordulhatnak azonban olyan esetek, amikor az LLM-ek nem tudják teljesen megoldani ezt a problémát tökéletesen, ha néhány kulcsfontosságú információ hiányzik. Ezért mindig ellenőrizze az eredményeket.**

**Íme néhány más felhasználási terület az adattisztításhoz AI-val:**

* **Duplikátumok eltávolítása:** A duplikált bejegyzések azonosítása és eltávolítása az adatkészletben.
* **Hiányzó értékek kezelése:** A hiányzó értékek kezelésének módjának meghatározása, akár imputálással, akár törléssel.
* **Adatformátumok szabványosítása:** Az adatmezők következetes formázásának biztosítása, mint például a dátumok, telefonszámok vagy címek.
* **Speciális karakterek eltávolítása:** A felesleges speciális karakterek vagy szimbólumok eltávolítása, amelyek befolyásolhatják az adatok minőségét.
* **Adattípusok konvertálása:** Az adatmezők megfelelő adattípusokra történő konvertálása (pl. karakterláncok számokká alakítása).
* **Outlierek (kiugró értékek) kezelése:** A kiugró értékek azonosítása és kezelése, amelyek torzíthatják az elemzést vagy a modellezési eredményeket
* **Nem következetes írásmódok feloldása:** Ugyanazon entitás inkonzisztens írásmódjainak vagy variációinak szabványosítása és korrigálása.
* **E-mail címek ellenőrzése:** E-mail címek érvényességének ellenőrzése és a formátumhibák javítása.
* **Szöveges adatok normalizálása:** A felesleges szóközök eltávolítása, kisbetűs átalakítás és az írásjelek kezelése szövegmezőkben.

**Annak biztosítása után, hogy az adatok tiszták és megbízhatóak, a hangsúly áttevődik a mintaillesztési technikák kihasználására a koherens és kontextusérzékeny válaszok előállításához.**


#Mintaillesztés, rendszerezés, információ kinyerés

Mi a mintaillesztés? A folyamat, mely során strukturálatlan szöveget elemzünk, hogy releváns mintákat és információkat azonosítsunk és kinyerjünk.
Példa: Személyes adatok kinyerése strukturálatlan szövegből, mint név, kor, foglalkozás.
Hasznosítás: A kinyert, strukturált adatokat jobban lehet rendszerezni, elemezni és más rendszerekkel integrálni.


#Folyamat leírása

* **kép szkennelése, beolvasása valamilyen forrásból**
* **OCR segítségével képből szöveg készítése, eredmény tipikusan struktúrálatlan szöveg**
* **LLM segítségével a fontos információk kinyerése**
* **LLM segítségével valamilyen átadható adatformátum készítése pl. JSON, ...**
* **adatok átadása valamilyen más szolgáltatásnak**

#Példa

Az A vállalatnak van egy sofőr csapata, akik szállításokat végeznek különböző ügyfeleknek több városban. A sofőröknek jelenteniük kell az általuk felmerülő költségeket, például az üzemanyagot, az útdíjakat, a járműkarbantartást, az étkezést és a szállást. Jelenleg minden nap végén küldeniük kell egy e-mailt, amelyben részletezik az egyes kiadásokat, és csatolniuk kell a számlák vagy nyugták képeit. Ez időigényes és hibákkal terhelt.

Emellett az irodában valakinek el kell olvasnia minden egyes e-mailt, át kell másolnia az adatokat egy táblázatba vagy más rendszerbe, és a képeket egy meghajtó mappába vagy valami hasonlóba kell helyeznie. Ez a folyamat hibalehetőségekkel teli, időigényes, és bonyolítja az adatok összekapcsolását.

Ha például az e-mailekből származó képeket egy Drive mappába mentik, míg más mezőket egy táblázatba másolnak, nehéz lehet összekapcsolni, hogy melyik kép melyik rekordhoz vagy sorhoz tartozik a táblázatban. Nincs automatikus kapcsolat az elkülönített adatok között.



##Alkalmazás az OCR-technológiával:
A felhasználó képet (pl. számlát) tölt fel OCR-feldolgozásra
Mesterséges intelligencia rendezi a kinyerhető szöveget
Az adatokból például automatikusan táblázatba kerülnek a költségek

In [ ]:
!gdown 1GesvS7nhgHK3eke-t2RsQYGXGSVy_2g1 #adatok beszerzése
!gdown 1OcuK4c6UdoqGGSRPEn3Zxip8EfvhBXQU
!gdown 1RdBidgJM4RxdgKtRjGtgZ1TqwV7g5i1y

Downloading...
From: https://drive.google.com/uc?id=1GesvS7nhgHK3eke-t2RsQYGXGSVy_2g1
To: /content/Aldi.jpg
100% 253k/253k [00:00<00:00, 114MB/s]
Downloading...
From: https://drive.google.com/uc?id=1OcuK4c6UdoqGGSRPEn3Zxip8EfvhBXQU
To: /content/szamla1.jpg
100% 126k/126k [00:00<00:00, 15.0MB/s]
Downloading...
From: https://drive.google.com/uc?id=1RdBidgJM4RxdgKtRjGtgZ1TqwV7g5i1y
To: /content/szamla.jpg
100% 161k/161k [00:00<00:00, 61.4MB/s]


In [ ]:
!apt-get install tesseract-ocr
!pip install pytesseract pillow
!sudo apt-get install tesseract-ocr-hun -y

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
tesseract-ocr is already the newest version (4.1.1-2.1build1).
0 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  tesseract-ocr-hun
0 upgraded, 1 newly installed, 0 to remove and 2 not upgraded.
Need to get 1,853 kB of archives.
After this operation, 5,311 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 tesseract-ocr-hun all 1:4.00~git30-7274cfa-1.1 [1,853 kB]
Fetched 1,853 kB in 1s (1,543 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 1.)
debconf: falling back to frontend: Readline
debconf: unable to initialize front

In [ ]:
import requests
from PIL import Image
import pytesseract
from io import BytesIO

url = "https://www.infocentrum.hu/wp-content/uploads/help_images/elolegszamla_kep.png"

response = requests.get(url)
image_bytes = response.content

kep = Image.open(BytesIO(image_bytes))

Prompt: készíts egy python szkriptet colabhoz ami beolvas egy képet majd az tesseract segtítségével felismeri a szövegeket, táblázatokat és kiírja azokat.

In [ ]:
from PIL import Image
import pytesseract
import io

# Beolvasandó kép elérési útja
kep_eleresi_ut = "Aldi.jpg"

# Kép betöltése
kep = Image.open(kep_eleresi_ut)

# Szöveg kinyerése
szoveg = pytesseract.image_to_string(kep,lang="hun")

# Táblázat kinyerése
tabulat = pytesseract.image_to_data(kep, lang="hun", output_type=pytesseract.Output.DICT)

# Kiíratás
print("Szöveg:")
print(szoveg)

#print("Táblázat:")
#for sor in tabulat['text']:
#    print(sor)


Szöveg:
   
  
  
  
    
  
 

Adi Magyarország Élelmiszer Bt.
2051 Biatorbágy
Mészárosok útja 2.

102. sz, üzlet
7634 ker
Nyugati Ipari
ADÓSZÁM: 22934663- 3. 44

— WIN —

B00 ME Ek 1799 1.52

5 DB x 75 Ft/DB 375
(00 sza nö.olaj . 1299 399
Ü S; Hi 999
BŰN Félzs. túró 2509 449
500 Félzs. túró 2509 449
B00 Tejföl 207 4509 469
T 499
619
619
49 Ft/DB 1347
00. Ananász 9809 999
COD Ananász 9809 999
C00 Almapüré a 99
9 299
SZ 9 399
21 699
CO P 4 429
C00 Ananász 5809 944
C00 Zabpehely 9009 399
(00 ketchup 9999 Üz
Bo ki ő . ex. Sz, 790m! 2799
OVáSZOS CIPÓ
3 DB x 499 Ft/DB 1497
C00 Brokkoli 4509 649
BD Fr .camemb. 2909 4 180
an 8 tea 1do 299
urgonya Pp.
. 4107 1 Tára
728
1 459
1798
7699
21.269 Ft

 

21 263 Ft
21 263 Ft



In [ ]:
tabulat

{'level': [1,
  2,
  3,
  4,
  5,
  4,
  5,
  4,
  5,
  4,
  5,
  4,
  5,
  5,
  4,
  5,
  4,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  5,
  5,
  2,
  3,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  4,
  5,
  4,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
  5,
  5,
  5,
  4,
  5,
  5,
 

Prompt: A következő számla szövegben keresd meg szolgáltató adatait és a fizetendő összeget készíts belőle JSON objektumot:
szöveg:

  
    
  


Adi Magyarország Élelmiszer Bt.
2051 Biatorbágy
Mészárosok útja 2.

102. sz, üzlet
7634 ker
Nyugati Ipari
ADÓSZÁM: 22934663- 3. 44

— WIN —

B00 ME Ek 1799 1.52

5 DB x 75 Ft/DB 375
(00 sza nö.olaj . 1299 399
Ü S; Hi 999
BŰN Félzs. túró 2509 449
500 Félzs. túró 2509 449
B00 Tejföl 207 4509 469
T 499
619
619
49 Ft/DB 1347
00. Ananász 9809 999
COD Ananász 9809 999
C00 Almapüré a 99
9 299
SZ 9 399
21 699
CO P 4 429
C00 Ananász 5809 944
C00 Zabpehely 9009 399
(00 ketchup 9999 Üz
Bo ki ő . ex. Sz, 790m! 2799
OVáSZOS CIPÓ
3 DB x 499 Ft/DB 1497
C00 Brokkoli 4509 649
BD Fr .camemb. 2909 4 180
an 8 tea 1do 299
urgonya Pp.
. 4107 1 Tára
728
1 459
1798
7699
21.269 Ft



21 263 Ft
21 263 Ft


Kimenet:
{
  "szolgaltato": {
    "nev": "Adi Magyarország Élelmiszer Bt.",
    "cim": "2051 Biatorbágy, Mészárosok útja 2., 102. sz, üzlet, 7634 ker",
    "adoszam": "22934663-3-44"
  },
  "fizetendo_osszeg": 21263
}


In [ ]:
{
  "szolgaltato": {
    "nev": "Adi Magyarország Élelmiszer Bt.",
    "cim": "2051 Biatorbágy, Mészárosok útja 2., 102. sz, üzlet, 7634 ker",
    "adoszam": "22934663-3-44"
  },
  "fizetendo_osszeg": 21263
}


{'szolgaltato': {'nev': 'Adi Magyarország Élelmiszer Bt.',
  'cim': '2051 Biatorbágy, Mészárosok útja 2., 102. sz, üzlet, 7634 ker',
  'adoszam': '22934663-3-44'},
 'fizetendo_osszeg': 21263}

#Feladat:

![img](https://www.infocentrum.hu/wp-content/uploads/help_images/elolegszamla_kep.png)

Elad

In [ ]:
import requests
from io import BytesIO

url = "https://www.infocentrum.hu/wp-content/uploads/help_images/elolegszamla_kep.png"

response = requests.get(url)
image_bytes = response.content

kep = Image.open(BytesIO(image_bytes))

In [ ]:
szoveg = pytesseract.image_to_string(kep,lang="hun")

# Táblázat kinyerése
tabulat = pytesseract.image_to_data(kep, lang="hun", output_type=pytesseract.Output.DICT)

# Kiíratás
#print("Szöveg:")
#print(szoveg)

print("Táblázat:")
for sor in tabulat['text']:
    print(sor)

Táblázat:




 



infocentrum

szoftver
stúdió



Előleg
számla



Sorszám:
000007/2015



 



 



Szállító

Infocentrum
Szoftver
Stúdió

7400
Kaposvár,
Vár
u.
12-14.



Adószám:
12345678-1-12

Közösségi
adószám
HU12345678

Bankszámlaszám:

tel:06-30-916-8888


fax
821312-391

wwrwinfocentrumhu



Vevő

TesztKereskedelmi
Kft.



Adószám:
12121212-1-21

Közösségi
adószám
HU12121212

Bankszámlaszám:



 



1023
Budapest
II.
kerület,
Madár
u.
55.



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



 



Megrendelés
száma:
Fizetés
módja:
Telj.
időpontja:
Számla
kelte:
Fiz.
határidő;

Átutalás
2015.06.12.
2015.06.12.
2015.06.12.

Egyéb
adatok:

orolási
szám,
Megnevezés
e.
ÁT]
Átaérté

VTSZ/
Besorolási
szám,
Megi
ME]
Menny.
[
ágmélken
té]
Áraérték

ELŐLEG
ri
a]
50
000
50000]
2796
13500
63
500]

4
lioenszes
hálózatos
számlázó
praram
előlege.

Ji
dési
ök
tsdlé
Számlaérték:
50.000
Ft

Áfa:
13500
Ft

27
50000
Ft
13500Ft
63500Ft

Fizetendő:
$3500Ft



 

#Megoldás:

Prompt:






Kimenet:

{
  "számla_sorszám": "000007/2015",
  "szolgáltató": {
    "név": "Infocentrum Szoftver Stúdió",
    "cím": "7400 Kaposvár, Vár u. 12-14.",
    "adószám": "12345678-1-12",
    "közösségi_adószám": "HU12345678",
    "bankszámlaszám": "",
    "telefonszám": "06-30-916-8888",
    "fax": "821312-391",
    "weboldal": "www.infocentrum.hu"
  },
  "vevő": {
    "név": "TesztKereskedelmi Kft.",
    "cím": "1023 Budapest II. kerület, Madár u. 55.",
    "adószám": "12121212-1-21",
    "közösségi_adószám": "HU12121212",
    "bankszámlaszám": ""
  },
  "megvásárolt_termék": "4 licencses hálózatos számlázó program előlege",
  "mennyiség": 1,
  "ár": 50000,
  "áfa": 27,
  "számlaérték": 50000,
  "fizetendő": 63500,
  "fizetés_módja": "átutalás",
  "teljesítés_időpontja": "2015.06.12.",
  "számla_kelte": "2015.06.12.",
  "fizetési_határidő": "2015.06.12."
}


In [ ]:
{
  "számla_sorszám": "000007/2015",
  "szolgáltató": {
    "név": "Infocentrum Szoftver Stúdió",
    "cím": "7400 Kaposvár, Vár u. 12-14.",
    "adószám": "12345678-1-12",
    "közösségi_adószám": "HU12345678",
    "bankszámlaszám": "",
    "telefonszám": "06-30-916-8888",
    "fax": "821312-391",
    "weboldal": "www.infocentrum.hu"
  },
  "vevő": {
    "név": "TesztKereskedelmi Kft.",
    "cím": "1023 Budapest II. kerület, Madár u. 55.",
    "adószám": "12121212-1-21",
    "közösségi_adószám": "HU12121212",
    "bankszámlaszám": ""
  },
  "megvásárolt_termék": "4 licencses hálózatos számlázó program előlege",
  "mennyiség": 1,
  "ár": 50000,
  "áfa": 27,
  "számlaérték": 50000,
  "fizetendő": 63500,
  "fizetés_módja": "átutalás",
  "teljesítés_időpontja": "2015.06.12.",
  "számla_kelte": "2015.06.12.",
  "fizetési_határidő": "2015.06.12."
}


{'számla_sorszám': '000007/2015',
 'szolgáltató': {'név': 'Infocentrum Szoftver Stúdió',
  'cím': '7400 Kaposvár, Vár u. 12-14.',
  'adószám': '12345678-1-12',
  'közösségi_adószám': 'HU12345678',
  'bankszámlaszám': '',
  'telefonszám': '06-30-916-8888',
  'fax': '821312-391',
  'weboldal': 'www.infocentrum.hu'},
 'vevő': {'név': 'TesztKereskedelmi Kft.',
  'cím': '1023 Budapest II. kerület, Madár u. 55.',
  'adószám': '12121212-1-21',
  'közösségi_adószám': 'HU12121212',
  'bankszámlaszám': ''},
 'megvásárolt_termék': '4 licencses hálózatos számlázó program előlege',
 'mennyiség': 1,
 'ár': 50000,
 'áfa': 27,
 'számlaérték': 50000,
 'fizetendő': 63500,
 'fizetés_módja': 'átutalás',
 'teljesítés_időpontja': '2015.06.12.',
 'számla_kelte': '2015.06.12.',
 'fizetési_határidő': '2015.06.12.'}

## Gemini API integráció

Az eddigiekben manuálisan másoltuk az OCR szöveget a ChatGPT-be. Most **automatizáljuk az egész folyamatot** a Google Gemini API segítségével.

**A fejezet témái:**
* Gemini API beállítása Colab-ban
* Érzelem-elemzés futtatható kóddal
* Képelőfeldolgozás (preprocessing) a pontosabb OCR-ért
* Teljes számla-feldolgozó pipeline: Tesseract OCR + Gemini
* Modern megközelítés: Vision LLM (Tesseract nélkül)
* Kötegelt feldolgozás több számlán

In [ ]:
# Szükséges csomagok telepítése
!pip install google-generativeai --quiet
!pip install opencv-python-headless --quiet

### API kulcs beállítása

A Gemini API kulcsot a [Google AI Studio](https://aistudio.google.com/apikey) oldalon lehet ingyenesen igényelni.

**Colab-ban biztonságos tárolás:** bal oldali panel → kulcs ikon → `GEMINI_API_KEY` névvel felvenni.

In [ ]:
import google.generativeai as genai
from google.colab import userdata
import json

# API kulcs betöltése Colab Secrets-ből (biztonságos)
try:
    api_key = userdata.get('GEMINI_API_KEY')
    print("API kulcs sikeresen betöltve a Colab Secrets-ből.")
except Exception:
    import getpass
    api_key = getpass.getpass('Add meg a Gemini API kulcsot: ')

genai.configure(api_key=api_key)

# Modell inicializálása (gemini-1.5-flash: ingyenes tierben elérhető, gyors)
model = genai.GenerativeModel('gemini-2.5-flash')
print(f"Modell kész: {model.model_name}")

API kulcs sikeresen betöltve a Colab Secrets-ből.
Modell kész: models/gemini-2.5-flash


### Érzelem-elemzés Gemini API-val

Az előző fejezetben kézzel írtuk be a promptot a ChatGPT-be. Most ugyanezt **Python kódból** hívjuk meg.

In [ ]:
def sentiment_elemzes(szoveg: str) -> str:
    """Egyetlen mondat érzelem-elemzése. Visszaadja: pozitív / semleges / negatív"""
    prompt = f"""Azonosítsd és kategorizáld az alábbi szövegben kifejezett érzelmet.
Csak egy szóban válaszolj: pozitív, semleges, vagy negatív.

Szöveg: {szoveg}"""
    response = model.generate_content(prompt)
    return response.text.strip()

# Tesztelés
teszt_mondatok = [
    "Ez a termék csodálatos! Tényleg javított az életemen.",
    "A projektjavaslatom beadási határideje jövő pénteken 17 órakor lejár.",
    "Nagyon csalódott vagyok a szörnyű ügyfélszolgálat miatt.",
    "Még ma be kell ugranom az élelmiszerboltba tejért és tojásért.",
]

print(f"{'Szöveg (rövidítve)':<55} | {'Hangulat'}")
print("-" * 70)
for mondat in teszt_mondatok:
    eredmeny = sentiment_elemzes(mondat)
    print(f"{mondat[:52]+'...':<55} | {eredmeny}")

Szöveg (rövidítve)                                      | Hangulat
----------------------------------------------------------------------
Ez a termék csodálatos! Tényleg javított az életemen... | pozitív
A projektjavaslatom beadási határideje jövő pénteken... | Semleges
Nagyon csalódott vagyok a szörnyű ügyfélszolgálat mi... | negatív
Még ma be kell ugranom az élelmiszerboltba tejért és... | semleges


### Tömeges érzelem-elemzés – JSON kimenet

A Gemini `response_mime_type="application/json"` beállítással garantáltan valid JSON-t ad vissza, ami közvetlenül feldolgozható.

In [ ]:
mondatok = [
    "Nagyon csalódott vagyok a szörnyű ügyfélszolgálat miatt, amit kaptam.",
    "Az ételek ebben az étteremben abszolút finomak. Mindenképpen vissza fogok térni.",
    "Nemrég vettem ezt a laptopot, és ez a legjobb technikai vásárlás évek óta.",
    "Úgy néz ki, hogy ma délután esni fog. Vinnem kéne egy esernyőt.",
    "A projektjavaslatom beadási határideje jövő pénteken 17 órakor lejár.",
    "Most néztem meg az új Marvel filmet – teljesen felülmúlta az elvárásaimat.",
    "Ha manapság a híreket olvasom, csak szorongok és lehangolt vagyok.",
    "Ez a legrosszabb ízű étel, amit valaha ettem egy étteremben.",
]

szovegek_str = "\n".join([f'{i+1}. "{m}"' for i, m in enumerate(mondatok)])

prompt = f"""Azonosítsd és kategorizáld az alábbi mondatokban kifejezett érzelmeket.
Csak 'pozitív', 'semleges' vagy 'negatív' értékeket használj.
Válaszolj JSON formátumban: egy "eredmenyek" kulcsú lista, ahol minden elem tartalmaz
"sorszam", "szoveg" és "hangulat" mezőket.

Mondatok:
{szovegek_str}"""

response = model.generate_content(
    prompt,
    generation_config=genai.GenerationConfig(
        response_mime_type="application/json"
    )
)

eredmenyek = json.loads(response.text)["eredmenyek"]

# Táblázatos megjelenítés
print(f"{'#':<4} | {'Hangulat':<12} | Szöveg")
print("-" * 80)
for e in eredmenyek:
    print(f"{e['sorszam']:<4} | {e['hangulat']:<12} | {e['szoveg'][:55]}...")

#    | Hangulat     | Szöveg
--------------------------------------------------------------------------------
1    | negatív      | Nagyon csalódott vagyok a szörnyű ügyfélszolgálat miatt...
2    | pozitív      | Az ételek ebben az étteremben abszolút finomak. Mindenk...
3    | pozitív      | Nemrég vettem ezt a laptopot, és ez a legjobb technikai...
4    | semleges     | Úgy néz ki, hogy ma délután esni fog. Vinnem kéne egy e...
5    | semleges     | A projektjavaslatom beadási határideje jövő pénteken 17...
6    | pozitív      | Most néztem meg az új Marvel filmet – teljesen felülmúl...
7    | negatív      | Ha manapság a híreket olvasom, csak szorongok és lehang...
8    | negatív      | Ez a legrosszabb ízű étel, amit valaha ettem egy éttere...


In [ ]:
eredmenyek

[{'sorszam': 1,
  'szoveg': 'Nagyon csalódott vagyok a szörnyű ügyfélszolgálat miatt, amit kaptam.',
  'hangulat': 'negatív'},
 {'sorszam': 2,
  'szoveg': 'Az ételek ebben az étteremben abszolút finomak. Mindenképpen vissza fogok térni.',
  'hangulat': 'pozitív'},
 {'sorszam': 3,
  'szoveg': 'Nemrég vettem ezt a laptopot, és ez a legjobb technikai vásárlás évek óta.',
  'hangulat': 'pozitív'},
 {'sorszam': 4,
  'szoveg': 'Úgy néz ki, hogy ma délután esni fog. Vinnem kéne egy esernyőt.',
  'hangulat': 'semleges'},
 {'sorszam': 5,
  'szoveg': 'A projektjavaslatom beadási határideje jövő pénteken 17 órakor lejár.',
  'hangulat': 'semleges'},
 {'sorszam': 6,
  'szoveg': 'Most néztem meg az új Marvel filmet – teljesen felülmúlta az elvárásaimat.',
  'hangulat': 'pozitív'},
 {'sorszam': 7,
  'szoveg': 'Ha manapság a híreket olvasom, csak szorongok és lehangolt vagyok.',
  'hangulat': 'negatív'},
 {'sorszam': 8,
  'szoveg': 'Ez a legrosszabb ízű étel, amit valaha ettem egy étteremben.',
  'ha

### Teljes pipeline: Tesseract OCR + Gemini

A folyamat lépései kódban:
1. Képelőfeldolgozás (OpenCV) (ezt kihagyjuk)
2. Szöveg kinyerése (Tesseract)
3. Strukturálás (Gemini) → JSON

In [ ]:
def szamla_feldolgozas_ocr_gemini(kep_ut: str) -> dict:
    """
    Teljes számla-feldolgozó pipeline:
      Kép → OpenCV előfeldolgozás → Tesseract OCR → Gemini → JSON
    """
    # 1. lépés: képelőfeldolgozás + OCR
    ocr_szoveg = pytesseract.image_to_string(kep_ut, lang="hun")
    print(f"[1/2] OCR kész – {len(ocr_szoveg)} karakter kinyerve")

    # 2. lépés: Gemini strukturálja az adatokat
    prompt = f"""Az alábbi szöveg egy OCR-rel beolvasott számláról/nyugtáról származik,
ezért tartalmazhat helyesírási hibákat és zajt.

Keresd meg és nyerd ki a következő adatokat JSON formátumban:
- szallito: nev, cim, adoszam
- vevo: nev, cim, adoszam  (ha nincs, értéke null)
- szamla_sorszam            (ha nincs, értéke null)
- szamla_kelte              (ha nincs, értéke null)
- tetelek: lista – minden elemben: megnevezes, mennyiseg, egysegar, osszeg
- fizetendo_osszeg          (csak szám, Ft jel nélkül)
- fizetesi_mod              (ha nincs, értéke null)

Ha valamit nem találsz, az értéke legyen null. Csak valid JSON-t adj vissza!

OCR szöveg:
{ocr_szoveg}"""

    response = model.generate_content(
        prompt,
        generation_config=genai.GenerationConfig(
            response_mime_type="application/json"
        )
    )
    print("[2/2] Gemini strukturálás kész")
    return json.loads(response.text)

# Teszt
eredmeny = szamla_feldolgozas_ocr_gemini("Aldi.jpg")
print("\n--- Kinyert adatok ---")
print(json.dumps(eredmeny, indent=2, ensure_ascii=False))

[1/2] OCR kész – 829 karakter kinyerve
[2/2] Gemini strukturálás kész

--- Kinyert adatok ---
{
  "szallito": {
    "nev": "Adi Magyarország Élelmiszer Bt.",
    "cim": "2051 Biatorbágy Mészárosok útja 2.",
    "adoszam": "22234683-9-44"
  },
  "vevo": null,
  "szamla_sorszam": null,
  "szamla_kelte": null,
  "tetelek": [
    {
      "megnevezes": "hal tzjAB",
      "mennyiseg": null,
      "egysegar": 1799,
      "osszeg": 1.98
    },
    {
      "megnevezes": "Félzs. túró",
      "mennyiseg": 500,
      "egysegar": null,
      "osszeg": 449
    },
    {
      "megnevezes": "Tejföl 20X",
      "mennyiseg": 45,
      "egysegar": null,
      "osszeg": 459
    },
    {
      "megnevezes": "Kej",
      "mennyiseg": null,
      "egysegar": null,
      "osszeg": 819
    },
    {
      "megnevezes": "Céklalé",
      "mennyiseg": 11,
      "egysegar": null,
      "osszeg": 619
    },
    {
      "megnevezes": "len gyümölcslé",
      "mennyiseg": null,
      "egysegar": 9,
      "osszeg": 1347

### Vision LLM megközelítés – Tesseract nélkül

A **Gemini 2.5 Flash** képes közvetlenül képből olvasni – nem kell Tesseract.
Ez pontosabb, gyorsabb, és kevesebb kód.

**Mikor melyiket érdemes használni?**

| Megközelítés | Előny | Hátrány |
|---|---|---|
| Tesseract + Gemini | Ingyenes OCR, offline is futtatható | Zajos OCR ront a minőségen |
| Gemini Vision | Pontosabb, egyszerűbb kód | API hívás száma korlátozott |

In [ ]:
def szamla_feldolgozas_vision(kep_ut: str) -> dict:
    """
    Vision LLM megközelítés: a kép közvetlenül megy a Gemini-be, Tesseract nélkül.
    """
    kep = Image.open(kep_ut)

    prompt = """Olvasd be ezt a számlát vagy nyugtát és nyerd ki az adatokat JSON formátumban:
- szallito: nev, cim, adoszam
- vevo: nev, cim, adoszam  (ha nincs, értéke null)
- szamla_sorszam            (ha nincs, értéke null)
- szamla_kelte              (ha nincs, értéke null)
- tetelek: lista – minden elemben: megnevezes, mennyiseg, egysegar, osszeg
- fizetendo_osszeg          (csak szám, Ft jel nélkül)
- fizetesi_mod              (ha nincs, értéke null)

Ha valamit nem találsz, az értéke legyen null. Csak valid JSON-t adj vissza!"""

    response = model.generate_content(
        [prompt, kep],
        generation_config=genai.GenerationConfig(
            response_mime_type="application/json"
        )
    )
    return json.loads(response.text)

# Összehasonlítás a két megközelítés között
print("=" * 50)
print("Tesseract + Gemini pipeline:")
print("=" * 50)
eredmeny_ocr = szamla_feldolgozas_ocr_gemini("Aldi.jpg")
print(json.dumps(eredmeny_ocr, indent=2, ensure_ascii=False))

print("\n" + "=" * 50)
print("Gemini Vision (közvetlen):")
print("=" * 50)
eredmeny_vision = szamla_feldolgozas_vision("Aldi.jpg")
print(json.dumps(eredmeny_vision, indent=2, ensure_ascii=False))

Tesseract + Gemini pipeline:
[1/2] OCR kész – 829 karakter kinyerve
[2/2] Gemini strukturálás kész
{
  "szallito": {
    "nev": "Adi Magyarország Élelmiszer Bt.",
    "cim": "2051 Biatorbágy\nMészárosok útja 2.",
    "adoszam": "22234683- 9. 44"
  },
  "vevo": {
    "nev": null,
    "cim": null,
    "adoszam": null
  },
  "szamla_sorszam": null,
  "szamla_kelte": null,
  "tetelek": [
    {
      "megnevezes": "hal tzjAB",
      "mennyiseg": 1.98,
      "egysegar": 1799,
      "osszeg": 3562.02
    },
    {
      "megnevezes": null,
      "mennyiseg": 9,
      "egysegar": 75,
      "osszeg": 675
    },
    {
      "megnevezes": "Félzs. túró",
      "mennyiseg": 500,
      "egysegar": null,
      "osszeg": 449
    },
    {
      "megnevezes": "Tejföl 20X",
      "mennyiseg": 45,
      "egysegar": null,
      "osszeg": 459
    },
    {
      "megnevezes": "Kei]",
      "mennyiseg": 1,
      "egysegar": null,
      "osszeg": 819
    },
    {
      "megnevezes": "Céklalé",
      "mennyiseg"

### Kötegelt feldolgozás – több számla egyszerre

A sofőrök naponta több számlát küldenek be. Az alábbi kód **automatikusan feldolgoz minden képet** és összesíti az eredményeket egy táblázatba.

In [ ]:
import pandas as pd

def kotegelt_feldolgozas(kep_lista: list, modszer: str = "vision") -> pd.DataFrame:
    """
    Több számla/nyugta feldolgozása és összesítése DataFrame-be.
    modszer: 'vision' (Gemini Vision) vagy 'ocr' (Tesseract + Gemini)
    """
    feldolgozo = szamla_feldolgozas_vision if modszer == "vision" else szamla_feldolgozas_ocr_gemini
    eredmenyek = []

    for kep_ut in kep_lista:
        print(f"Feldolgozás: {kep_ut} ...", end=" ")
        try:
            adat = feldolgozo(kep_ut)
            adat["forras_fajl"] = kep_ut
            eredmenyek.append(adat)
            print("OK")
        except Exception as e:
            print(f"HIBA: {e}")
            eredmenyek.append({"forras_fajl": kep_ut, "hiba": str(e)})

    return pd.DataFrame(eredmenyek)

# Összes letöltött számla feldolgozása
kepek = ["Aldi.jpg", "szamla1.jpg", "szamla.jpg"]
df = kotegelt_feldolgozas(kepek, modszer="vision")

# Összesített nézet
print("\n--- Összesítő táblázat ---")
oszlopok = [c for c in ["forras_fajl", "szamla_sorszam", "szamla_kelte", "fizetendo_osszeg"] if c in df.columns]
print(df[oszlopok].to_markdown(index=False))

# Mentés CSV-be
df.to_csv("szamlak_osszesito.csv", index=False, encoding="utf-8-sig")
print("\nEredmény elmentve: szamlak_osszesito.csv")

Feldolgozás: Aldi.jpg ... OK
Feldolgozás: szamla1.jpg ... OK
Feldolgozás: szamla.jpg ... OK

--- Összesítő táblázat ---
| forras_fajl   |   szamla_sorszam | szamla_kelte   |   fizetendo_osszeg |
|:--------------|-----------------:|:---------------|-------------------:|
| Aldi.jpg      |                  | 2023.01.30     |            21263   |
| szamla1.jpg   |        303301754 | 2024.01.15     |             6010   |
| szamla.jpg    |          0003521 | 6/5/2016       |             3367.2 |

Eredmény elmentve: szamlak_osszesito.csv
